sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -446,7 +451,12 @@ case class DataSource(
         dataSource.createRelation(
           sparkSession.sqlContext, mode, caseInsensitiveOptions, Dataset.ofRows(sparkSession, data))
       case format: FileFormat =>
-        sparkSession.sessionState.executePlan(planForWritingFileFormat(format, mode, data)).toRdd
+        val qe = sparkSession.sessionState.executePlan(planForWritingFileFormat(format, mode, data))
+        qe.executedPlan.transform {
+          case f: FileWritingCommandExec =>
+            val newCmd = f.cmd.withExternalMetrics(externalMetrics.getOrElse(null))
+            FileWritingCommandExec(newCmd, f.children)
+        }.execute()
```

Comment:
Can you provide specified suggestion regarding this? I don't think there is a possible way to avoid this.

Those commands are invoking another commands to write the data out. We want to have the metrics updated is the metrics in the original physical node, not the invoked one. To do this, we either pass the callback or the metrics in the original plan into the invoked one. Do you think we have other choice?